orttraining/orttraining/models/runner/pipeline.cc

```diff
@@ -0,0 +1,286 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/models/runner/pipeline.h"
+
+#include <iostream>
+#include <vector>
+#include <cstdint>
+#include <cstdlib>
+#include <stdexcept>
+#include <thread>
+
+#include "gsl/gsl"
+#include "core/framework/ml_value.h"
+#include "orttraining/training_ops/cpu/controlflow/event_pool.h"
+
+namespace onnxruntime {
+namespace training {
+namespace pipeline {
+
+
+Slot::Slot() {
+    type = Empty;
+    batch_id = 0;
+}
+
+bool Slot::IsEmpty() const {
+    return type == Empty;
+}
+
+bool Slot::IsForward() const {
+    return type == Forward;
+}
+
+bool Slot::IsBackward() const {
+    return type == Backward;
+}
+
+void Slot::show() const {
+    int wid = waited_events.size() > 0 ? waited_events[0] : -1;
+    int rid = recorded_events.size() > 0 ? recorded_events[0] : -1;
+    switch (type) {
+    case Empty:
+        std::cout << "         ";
+        break;
+    case Forward:
+        std::cout << " F" << batch_id << "(" << wid << "," << rid << ")" << " ";
+        break;
+    case Backward:
+        std::cout << " B" << batch_id << "(" << wid << "," << rid << ")" << " ";
+        break;
+    }
+}
+
+PipelineSchedule::PipelineSchedule(int num_stages) {
```

Comment:
Main code for scheduling pipeline computation. A large part of it is borrowed from Ke's code, but due to some bugs identified during development, it's largely changed.